### PR TITLE
Add json rpc spec to searcher examples

### PR DESCRIPTION
**Pull request by ebin-mathews**
Moved to https://github.com/jito-labs/mev-protos/pull/36

### CODE REVIEW

File: http.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+### Response
+
+```json
+{
+    "jsonrpc": "2.0",
+    "result": {
+        "bundle_id": "3Eq21vXNB5s86c62bVuUfTeaMif1N2kUqRPBmGRJhyTA",
+        "meta": {
+            "Accepted": {
```
Comment (ebin-mathews) [inline http.md:246]: For finalized, currently its an empty message showing the bundle was successful. Do we need to add the slots the transactions were placed? (We don't do that in the current grpc api).

File: http.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+### Response
+
+```json
+{
+    "jsonrpc": "2.0",
+    "result": {
+        "bundle_id": "3Eq21vXNB5s86c62bVuUfTeaMif1N2kUqRPBmGRJhyTA",
+        "meta": {
+            "accepted": {
```
Comment (ebin-mathews) [inline http.md:246]: For finalized, currently its an empty message showing the bundle was successful. Do we need to add the slots the transactions were placed? (We don't do that in the current grpc api).

File: http.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+#### Request
+
+```bash
+curl http://localhost:8899 -X POST -H "Content-Type: application/json" -d '
+{
+    "jsonrpc": "2.0",
+    "id": 1,
+    "method": "sendBundle",
+    "params": [
```
Comment (buffalu) [inline http.md:165]: we might want to nest the bundles so you can attach other context to this like commitment, etc.

should look to see how other RPC endpoints handle nested list

File: http.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+}
+```
+
+Sample error codes:
+
+```json
+{
+    "results": [
+        {
+            "accepted": {
```
Comment (buffalu) [inline http.md:262]: i don't think it makes sense to expose this. returning this is a bug i think

File: http.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ebin-mathews) [reply http.md:246]: Would "forwarded" make more sense?

File: http.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+Returns bundle details for a submitted bundle
+
+### Parameters
+
+- `bundle_id`: `string`  a base-58/64 encoded string (required).
+- `object` (optional)
+  - Configuration object containing the following fields (optional):
+    - `commitment`: String. [Configure state commitment](#configuring-state-commitment).
+    - `encoding`: String (Default: "json"). Encoding for the returned Transaction. Possible values: json, base64, base58.
```
Comment (ebin-mathews) [inline http.md:195]: maybe limit to base64 everywhere?

File: http.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
+# JSON RPC HTTP Methods
+
+searcher nodes accept HTTP requests using the [JSON-RPC 2.0](https://www.jsonrpc.org/specification) specification.
+
+## RPC HTTP Endpoint
+
+**Default port:** 8899 e.g. [http://localhost:8899](http://localhost:8899), [http://192.168.1.88:8899](http://192.168.1.88:8899)
```
Comment (ebin-mathews) [inline http.md:7]: Is the port 8899 fine or should define something else?

File: http.md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+  recognized this block as finalized
+- `"confirmed"` - the node will query the most recent block that has been voted on by supermajority of the cluster.
+  - It incorporates votes from gossip and replay.
+  - It does not count votes on descendants of a block, only direct votes on that block.
+  - This confirmation level also upholds "optimistic confirmation" guarantees in  
+    release 1.3 and onwards.
+- `"processed"` - the node will query its most recent block. Note that the block  
+  may still be skipped by the cluster.
+
+For processing many dependent transactions in series, it's recommended to use
```
Comment (ebin-mathews) [inline http.md:63]: Add the custom commitment level definition

File: http.md [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+```
+
+### Response
+
+```json
+{
+    "jsonrpc": "2.0",
+    "result": {
+        "bundle_id": "3Eq21vXNB5s86c62bVuUfTeaMif1N2kUqRPBmGRJhyTA",
+        "slot": 1234,
```
Comment (ebin-mathews) [inline http.md:252]: What transactions were processed in different slots?

File: http.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        "meta": {
```
Comment (ebin-mathews) [inline http.md:253]: What other bundle details need to be exposed? Or are the states below sufficient?

File: http.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
+# JSON RPC HTTP Methods
+
+searcher nodes accept HTTP requests using the [JSON-RPC 2.0](https://www.jsonrpc.org/specification) specification.
```
Comment (esemeniuc) [inline http.md:3]: searcher node -> block engine?

File: http.md (hunk 6 above)
Comment (esemeniuc) [reply http.md:7]: 8899 is the default rpc port. it is also configurable

File: http.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+## RPC HTTP Endpoint
+
+**Default port:** 8899 e.g. [http://localhost:8899](http://localhost:8899), [http://192.168.1.88:8899](http://192.168.1.88:8899)
+
+## Request Formatting
+
+To make a JSON-RPC request, send an HTTP POST request with a `Content-Type: application/json` header. The JSON request data should contain 4 fields:
```
Comment (esemeniuc) [inline http.md:11]: maybe just link to solana docs instead? https://docs.solana.com/api/http#request-formatting

File: http.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+}
+
+```
+
+## getBundlesForAddress
+Returns signatures for confirmed bundles that include the given address in their accountKeys list. Returns signatures backwards in time from the provided signature or the most recent confirmed block.
+
+### Parameters:
+
+`<string>`: `required` Account address as a base-64 encoded string.
```
Comment (buffalu) [inline http.md:292]: addresses are typically base58... yeah i know its confusing lol

File: http.md (hunk 6 above)
Comment (buffalu) [reply http.md:7]: i think it'll be port 80/443

File: http.md [hunk 13]
```diff
@@ -0,0 +1,342 @@
+# JSON RPC HTTP Methods
+
+searcher nodes accept HTTP requests using the [JSON-RPC 2.0](https://www.jsonrpc.org/specification) specification.
+
+## RPC HTTP Endpoint
+
+**Default port:** 8899 e.g. [http://localhost:8899](http://localhost:8899), [http://192.168.1.88:8899](http://192.168.1.88:8899)
```
Comment (buffalu) [inline http.md:7]: cc @jedleggett wrt the proxy rework being done right now.

File: http.md [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+- `result: <array|number|object|string>` - requested data or success confirmation
+
+Requests can be sent in batches by sending an array of JSON-RPC request objects as the data for a single POST.
+
+## Definitions
+
+- Bundle: Bundles are a list of transactions that execute sequentially and atomically. “All or nothing” so to speak. This means that a user can send a bundle that contains multiple transactions and guarantee that they are all executed one after the other and the bundle succeeds only if all individual transactions succeed.
+- Hash: A SHA-256 hash of a chunk of data.
+- Pubkey: The public key of a Ed25519 key-pair.
+- Tip Account: List of accounts to which a tip can be paid for processing the bundles
```
Comment (esemeniuc) [inline http.md:37]: should mention that searcher pays tip

File: http.md [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+
+If commitment configuration is not provided, the node will default to `"finalized"` commitment
+
+Only methods that query bank state accept the commitment parameter. They are indicated in the API Reference below.
+
+#### RpcResponse Structure
+
+Many methods that take a commitment parameter return an RpcResponse JSON object comprised of two parts:
+
+- `context` : An RpcResponseContext JSON structure including a `slot` field at which the operation was evaluated.
```
Comment (esemeniuc) [inline http.md:85]: Link to RpcResponseContext structure or docs.rs?

File: http.md [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+
+#### Default
+
+If commitment configuration is not provided, the node will default to `"finalized"` commitment
+
+Only methods that query bank state accept the commitment parameter. They are indicated in the API Reference below.
+
+#### RpcResponse Structure
+
+Many methods that take a commitment parameter return an RpcResponse JSON object comprised of two parts:
```
Comment (esemeniuc) [inline http.md:83]: link to docs.rs would be helpful

File: http.md [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+        "E2eSqe33tuhAHKTrwky5uEjaVqnb2T9ns6nHHUrN8588",
+        "ARTtviJkLLt6cHGQDydfo1Wyk6M4VGZdKZ2ZhdnJL336"
+    ],
+    "id": 1
+}
+```
+
+## sendBundle
+
+Submits a bundled list of signed transaction to the cluster for processing. The transactions will be processed in order and will be atomic meaning if at least one of the transactions fail, the bundle won’t be processed. It’s all or nothing. This method does not alter the transaction in any way; it relays the transaction created by clients to the node as-is. If the node's rpc service accepts the bundle, this method immediately succeeds, without waiting for any confirmations. A successful response from this method does not guarantee the transaction is processed or confirmed by the cluster.
```
Comment (esemeniuc) [inline http.md:125]: might be simpler (not sure if incorrect)
```suggestion
Submits a bundled list of signed transaction to the cluster for processing. The transactions will be atomically processed in order, meaning if any of the transactions fail, the entire bundle won’t be processed (all or nothing). This method does not alter the transaction in any way; it relays the transaction created by clients to the node as-is. If the node's rpc service accepts the bundle, this method immediately returns a success response acknowledging that the bundle has been received. This does not guarantee the transaction is processed or landed on-chain.
```

File: http.md [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+
+### Parameters
+
+`<array[string]>`: `required` Fully-signed Transaction, as encoded string (base64).
+
+### Result
+
+The result field will be a JSON object with the following fields:
+
+- `result`: `<string>` - A bundle id which is the ash of all the first signatures of the transactions in the bundle, which is used to identify the bundle as a base-58/64 encoded `string`
```
Comment (esemeniuc) [inline http.md:135]:
```suggestion
- `result`: `<string>` - A bundle id which is the hash of all the first signatures of the transactions in the bundle, which is used to identify the bundle as a base-58/64 encoded `string`
```

File: http.md [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+  - `bundle`: `<object>` Bundle object in JSON format
+    - signatures: `<array[string]>` - A list of base-64 encoded signatures applied by the bundle. The list will not be empty. The first one is used as the bundle id. The rest are the transaction ids which are part of the bundle.
+  - `slot`: `<u64>`  The slot this bundle was processed in.
+  - `confirmationStatus`: `<string|null>` - The bundle transaction's cluster confirmation status; Either processed, confirmed, or finalized. See [Commitment](#configuring-state-commitment) for more on optimistic confirmation.
+  - `meta`: `<object|null>`: Transaction status metadata object:
+    - `err`: `<object|null>`: Error if the bundle failed, null if the bundle succeeded/finalized. The error reasons could be a bundle acceptance failure like not enough tips or failure of one the included transactions, for example blockhash expired. The transaction failure errors can be found at [Transaction errors](https://docs.rs/solana-sdk/latest/solana_sdk/transaction/enum.TransactionError.html#). If a transaction error, the transaction id will also be included to identify which transaction caused the failure.
```
Comment (esemeniuc) [inline http.md:194]:
```suggestion
    - `err`: `<object|null>`: Error if the bundle failed, null if the bundle succeeded/finalized. The error reasons could be a bundle acceptance failure (eg. not enough tips) or failure of one the included transactions (eg.  blockhash expired). The transaction failure errors can be found at [Transaction errors](https://docs.rs/solana-sdk/latest/solana_sdk/transaction/enum.TransactionError.html#). If a transaction error occurred, the transaction id will also be included to identify which transaction caused the failure.
```

File: http.md (hunk 12 above)
Comment (ebin-mathews) [reply http.md:274]: Is bundle_id base-58 as well? transaction id is base-58

File: http.md [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+                "message": "Bundle failed: Not enough tips"
+            }
+        },
+    ]
+}
+
+```
+
+## getBundlesForAddress
+Returns signatures for confirmed bundles that include the given address in their accountKeys list. Returns signatures backwards in time from the provided signature or the most recent confirmed block.
```
Comment (esemeniuc) [inline http.md:270]: Can you clarify the `backwards in time` bit? Is it sorted as oldest -> newest

File: http.md [hunk 21]
```diff
@@ -0,0 +1,324 @@
...
+
+### Parameters:
+
+`<string>`: `required` Account address as a base-64 encoded string.
+
+`<object>`: `optional` Configuration object containing the following fields:
+
+- `commitment`: `<string>`, optional
+- `limit`: `<u16>` optional, default: 100: Maximum transaction signatures to return (between 1 and 1,000).
```
Comment (esemeniuc) [inline http.md:279]: inclusive/exclusive?

File: http.md (hunk 20 above)
Comment (esemeniuc) [reply http.md:270]: seems like reverse-chronological order from below

File: http.md [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+
+### Result
+
+An array of `<object>`, ordered from the newest to the oldest transaction, containing transaction signature information with the following fields:
+
+- `signature`: `<string>` Transaction signature as a base-64 encoded string.
+- `slot`: `<u64>` The slot that contains the block with the transaction.
+- `err`: `<object|null>` Error if the bundle failed, null if the bundle succeeded. See [TransactionError definitions]((https://docs.rs/solana-sdk/latest/solana_sdk/transaction/enum.TransactionError.html#)) for more information if the included transactions failed. Otherwise if Bundle processing failed for other reasons, that would be provided.
+- `confirmationStatus`: `<string|null>` The bundle transaction's cluster confirmation status; Either processed, confirmed, or finalized. See Commitment for more on optimistic confirmation.
```
Comment (esemeniuc) [inline http.md:288]: link to commitment levels

**Review by esemeniuc — CHANGES_REQUESTED** [review]

File: http.md (hunk 15 above)
Comment (ebin-mathews) [reply http.md:67]: This is the same as in https://docs.solana.com/api/http#rpcresponse-structure I used the notes relevant to us.
It's mostly telling that in some cases, we provide the context that provide a slot field and whatever value that is returned. Sort of like

`### Response

```json
{
    "jsonrpc": "2.0",
    "result": {
        "context": {
            "slot": 82
        },
        "value": [
            {
                "bundle": {
                    "signatures": [
                        "3Eq21vXNB5s86c62bVuUfTeaMif1N2kUqRPBmGRJhyTA",
                        "2nBhEBYYvfaAe16UMNqRHre4YNSskvuYgx3M6E4JP1oDYvZEJHvoPzyUidNgNX5r9sTyN1J9UxtbCXy2rqYcuyuv"
                    ]
                },
                "slot": 1234,
                "confirmationStatus": "finalized",
                "err": "null"
            }
        ]
    },
    "id": 1
}`

**Pull request closed without merging**